File: samples/java/DeliveryDispatch/Server/CourierSpotNode/src/main/java/systems/zlink/samples/deliverydispatch/server/courierspotnode/ActorDirectory.java
package systems.zlink.samples.deliverydispatch.server.courierspotnode;

import java.util.HashMap;
import java.util.Map;

public final class ActorDirectory {
    private final Object gate = new Object();
    private final Map<String, CourierActor> actors = new HashMap<>();

    public void register(CourierActor actor) {
        synchronized (gate) {
            actors.put(actor.actorId(), actor);
        }
    }

    public CourierActor require(String actorId) {
        synchronized (gate) {
            CourierActor actor = actors.get(actorId);
            if (actor != null) {
                return actor;
            }
        }
        throw new IllegalStateException("Courier actor is not registered: " + actorId);
    }

    public void remove(String actorId) {
        synchronized (gate) {
            actors.remove(actorId);
        }
    }
}

File: samples/kotlin/DeliveryDispatch/Server/CourierSpotNode/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/courierspotnode/ActorDirectory.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode

class ActorDirectory {
    private val actors = mutableMapOf<String, CourierActor>()

    fun register(actor: CourierActor) {
        synchronized(actors) { actors[actor.actorId()] = actor }
    }

    fun require(actorId: String): CourierActor =
        synchronized(actors) { actors[actorId] }
            ?: throw IllegalStateException("Courier actor is not registered: $actorId")

    fun remove(actorId: String) {
        synchronized(actors) { actors.remove(actorId) }
    }
}
